Sources/LanguageServerProtocolTransport/JSONRPCConnection.swift

```diff
@@ -551,21 +435,22 @@ public final class JSONRPCConnection: Connection {
   /// - Important: Must be called on `queue`
   private func send(data dispatchData: DispatchData) {
     dispatchPrecondition(condition: .onQueue(queue))
+
     guard readyToSend() else { return }
 
     orLog("Writing output mirror file") {
-      try outputMirrorFile?.write(contentsOf: dispatchData)
+      try sendMirrorFile?.write(contentsOf: dispatchData)
     }
-    sendIO.write(offset: 0, data: dispatchData, queue: sendQueue) { [weak self] done, _, errorCode in
-      if errorCode != 0 {
-        logger.fault("IO error sending message \(errorCode)")
-        if done, let self {
-          // An unrecoverable error occurs on the channel’s file descriptor.
-          // Close the connection.
-          self.queue.async {
-            self.closeAssumingOnQueue()
-          }
-        }
+
+    do {
+      try self.sendFD.write(contentsOf: dispatchData)
+    } catch {
+      logger.fault("IO error sending message from \(self.name): \(error.forLogging)")
+      self.receiveFD.readabilityHandler = nil
+      // Match the pattern of `close()` but call `closeAssumingOnQueue` asynchronously to make sure that
+      // `closeAssumingOnQueue` is executed after all data from `receiveFD` has been read.
```

Comment:
I tried to remember why I added this code for quite a while and I think I’ve now remembered. Probably worth adding it in a comment so I don’t need to wonder about it again.


```suggestion
      // `closeAssumingOnQueue` is executed after all data from `receiveFD` has been read.
      // This is important in case `receiveFD` contains an error message that indicates why 
      // we might no longer be able to send data through the JSON-RPC connection.
```